### PR TITLE
Add chunked upload support

### LINKED ISSUE #123: Add support for chunked uploads in ddb share
As the title says.

### PR DESCRIPTION
**Pull request by HeDo88TH**
Closes #123

### CODE REVIEW

File: src/shareservice.cpp [hunk 1]
```diff
@@ -63,75 +74,93 @@ std::string ShareService::share(const std::vector<std::string> &input, const std
...
-                if (res.status() == 401){
-                    LOGD << "Token expired";
-                    authToken = reg.login(ac.username, ac.password);
-                    throw NetException("Unauthorized");
-                }
+        // TODO: Extract methods
+        if (filesize > maxUploadSize) {
+            // Upload in chunks
+            /*
+             * 1) Split file in manageable chunks
```
Comment (pierotofy) [inline src/shareservice.cpp:92]: Note: no need to split files, just read them from a specific byte offset. We don't want to write to the file system for this feature.

Comment (pierotofy) [reply src/shareservice.cpp:92]: See `mmap` and https://docs.microsoft.com/en-us/windows/win32/memory/creating-a-view-within-a-file

Comment (pierotofy) [reply src/shareservice.cpp:92]: Note this might require some research, as I've never done this

File: src/registry.cpp [hunk 2]
```diff
@@ -61,6 +61,8 @@ std::string Registry::login(const std::string &username,
         UserProfile::get()->getAuthManager()->saveCredentials(
             url, AuthCredentials(username, password));
 
+        this->authToken = token;
```
Comment (pierotofy) [inline src/registry.cpp:64]: Ok, be careful though.. long sessions (think long uploads from a rural area with dial-up speeds) could fail if the token expires. Login has now side effects.

Comment (pierotofy) [reply src/registry.cpp:64]: It was probably a problem before too, but now you have a state variable.

File: src/shareclient.cpp [hunk 3]
```diff
@@ -146,6 +147,96 @@ std::string ShareClient::Commit() {
...
+                    .authToken(this->registry->getAuthToken())
+                    .send();
+
+            if (res.status() != 200) this->registry->handleError(res);
+
+            json j = res.getJSON();
+
+            if (!j.contains("hash")) this->registry->handleError(res);
+
+            // Find a way to include this check
```
Comment (pierotofy) [inline src/shareclient.cpp:220]: This can probably be checked when the last chunk has been sent and registry has merged all the chunks.

(This one is kind of important to avoid corruption).

File: src/registry.cpp (hunk 2 above)
Comment (pierotofy) [reply src/registry.cpp:64]: A good check in `getAuthToken` would be to see if the token has expired, and if so, to re-do the login and return a refreshed token (now that we have the `expires` field...)

**Comment by pierotofy** [comment]
ETA for this one?

File: src/registry.cpp (hunk 2 above)
Comment (HeDo88TH) [reply src/registry.cpp:77]: Done!

**Comment by HeDo88TH** [comment]
Soon 👍

File: src/shareclient.cpp (hunk 3 above)
Comment (HeDo88TH) [reply src/shareclient.cpp]: Added in 2bcb9da

**Comment by HeDo88TH** [comment]
Unfortunately still not working. There is something wrong with the input stream I provide to `curl_mime_data_cb` 
I think I'm on the right track to fix it

**Comment by HeDo88TH** [comment]
It is now working on DroneDB side, I suppose I fixed all the issues (file open mode was not set binary by default and allocation type of carrier object). 
Unfortunately Registry is giving some strange errors when using streamed uploads (as `curl_mime_data_cb` does). Here it is a sample trace of what is going on:

```
2020-11-17 19:11:01.358 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65445)
2020-11-17 19:11:01.358 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.358 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 0
2020-11-17 19:11:01.358 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65445 data
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65445
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 65445
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 65445
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 130981
2020-11-17 19:11:01.359 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 130981
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 196517
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.360 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 196517
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 262053
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 262053
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.361 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 327589
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 327589
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 393125
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.362 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 393125
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 458661
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@169] curl_read_callback(1, 65536)
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@175] Index = 3
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@176] Current = 458661
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@177] Reading 65536 data
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@182] Read 65536
2020-11-17 19:11:01.363 DEBUG [21564] [ddb::net::Request::multiPartFormData::<lambda_a9dc478eeba1460b458d7ee4cfd62db3>::operator ()@183] Current = 524197
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Transfer-Encoding: chunked
< Content-Type: text/plain
< Server: Microsoft-IIS/10.0
< X-Powered-By: ASP.NET
< Date: Tue, 17 Nov 2020 18:11:00 GMT
* HTTP error before end of send, stop sending
<
* Closing connection 0
2020-11-17 19:11:01.544 DEBUG [21564] [ddb::Registry::handleError@122] Request error: System.IO.IOException: Unexpected end of Stream, the content may have already been read by another component.
   at Microsoft.AspNetCore.WebUtilities.MultipartReaderStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at System.IO.StreamReader.ReadBufferAsync(CancellationToken cancellationToken)
   at System.IO.StreamReader.ReadToEndAsyncInternal()
   at Microsoft.AspNetCore.WebUtilities.MultipartSectionStreamExtensions.ReadAsStringAsync(MultipartSection section)
   at Microsoft.AspNetCore.Http.Features.FormFeature.InnerReadFormAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.ModelBinding.FormFileValueProviderFactory.AddValueProviderAsync(ValueProviderFactoryContext context, HttpRequest request)
   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.CreateAsync(ActionContext actionContext, IList`1 factories)
   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.TryCreateAsync(ActionContext actionContext, IList`1 factories)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Registry.Web.Services.Adapters.TokenManagerMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in C:\Users\Art Director\source\repos\Registry\Registry.Web\Services\Adapters\TokenManagerMiddleware.cs:line 21
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

I did some testing and It seems to be the case that Registry tries to bind its parameters before the end of the upload stream. This leads to that ugly error. There is a way to disable the automatic binding (using a custom attribute `DisableFormValueModelBindingAttribute`) but I'm not sure it is the root of the error.

**Review by pierotofy — COMMENTED** [review]
This might still need some cleanup to make sure pointers are being deallocated, testing edge cases, etc. but the base case of uploading a file using chunks is working.

The problem with earlier code was that some assumptions were made about how read_callback and seek_callback work. It's always a very good idea to follow the structure of libcurl's examples and deviate as little as possible from them. It's quite likely that the stream was being repositioned at the wrong offset, or was reading the improper number of bytes, etc.

:+1: 

I removed the `filename` parameter, but perhaps that can be re-added if needed in the call to `curl_mime_filename`.

**Comment by HeDo88TH** [comment]
Great! I thought that the stream was able to keep its internal position consistent between callbacks but I was wrong 😭

File: src/net/request.cpp [hunk 4]
```diff
@@ -130,7 +130,7 @@ Request &Request::multiPartFormData(std::vector<std::string> files,
     // Add files
     for (unsigned long i = 0; i < files.size(); i += 2) {
         field = curl_mime_addpart(form);
-        curl_mime_filename(field, files[i].c_str());
+        curl_mime_name(field, files[i].c_str());
```
Comment (pierotofy) [inline src/net/request.cpp:133]: Why is this being changed?

**Comment by pierotofy** [comment]
Did you test progress reporting?

Does it work with multiple files?

If these all work, then we can merge.

**Comment by HeDo88TH** [comment]
Yes I tested it with both JPG and TXT files and it works well. I will go deeper when I will try the stress tests
